terraform/eks-cloudwatch/app_mesh.tf

```diff
@@ -0,0 +1,156 @@
+# so that we can shorten the execution time
+
+locals {
+  aws_account_id = var.aws_account_id != "" ? var.aws_account_id : data.aws_caller_identity.current.account_id
+  mesh_name = "howto-k8s-http-headers"
+}
+
+data "aws_caller_identity" "current" {}
+
+data "aws_iam_policy" "appmesh_policy" {
+  arn = "arn:aws:iam::${local.aws_account_id}:policy/AWSAppMeshK8sControllerIAMPolicy"
+}
+
+module "iam_assumable_role_with_oidc" {
+  source = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
+  version = "~> 3.0"
+  provider_url = data.aws_eks_cluster.testing_cluster.identity[0].oidc[0].issuer
+
+  create_role = true
+  role_name = "appmesh-controller_role"
+  role_policy_arns = [
+    data.aws_iam_policy.appmesh_policy.arn]
+
+  oidc_fully_qualified_subjects = [
+    "system:serviceaccount:${kubernetes_namespace.appmesh_ns.metadata[0].name}:appmesh-controller"]
+  number_of_role_policy_arns = 1
+}
+
+resource "kubernetes_namespace" "appmesh_ns" {
+  metadata {
+    name = "appmesh-system-${module.common.testing_id}"
+  }
+}
+
+resource "kubernetes_service_account" "appmesh_sa" {
+  metadata {
+    name = "appmesh-controller"
+    namespace = kubernetes_namespace.appmesh_ns.metadata[0].name
+    annotations = {
+      "eks.amazonaws.com/role-arn" = module.iam_assumable_role_with_oidc.this_iam_role_arn
+    }
+  }
+  automount_service_account_token = true
+}
+
+provider "helm" {
+  kubernetes {
+    config_path = "~/.kube/config"
+  }
+}
+
+resource "helm_release" "eks" {
+  name = "eks"
+  namespace = kubernetes_namespace.appmesh_ns.metadata[0].name
+
+  repository = "https://aws.github.io/eks-charts"
+  chart = "appmesh-controller"
+  version = "1.3.0"
+
+  set {
+    name = "serviceAccount.create"
+    value = "false"
+  }
+
+  set {
+    name = "serviceAccount.name"
+    value = kubernetes_service_account.appmesh_sa.metadata[0].name
+  }
+
+  set {
+    name = "region"
+    value = var.region
+  }
+
+  provisioner "local-exec" {
+    command = "kubectl apply -k \"github.com/aws/eks-charts/stable/appmesh-controller/crds?ref=v0.0.47\""
+  }
+
+  wait = true
+  depends_on = [module.iam_assumable_role_with_oidc]
+}
+
+resource "kubernetes_namespace" "traffic_ns" {
+  metadata {
+    name = "howto-k8s-http-headers-${module.common.testing_id}"
+    labels = {
+      "mesh" = local.mesh_name
+      "appmesh.k8s.aws/sidecarInjectorWebhook" = "enabled"
+    }
+  }
+  depends_on = [null_resource.delete_mesh]
+}
+
+resource "null_resource" "delete_mesh" {
+  provisioner "local-exec" {
+    when = destroy
+    command = "kubectl delete mesh howto-k8s-http-headers --ignore-not-found=true"
+  }
+  depends_on = [helm_release.eks]
+}
+
+resource "null_resource" "appmesh_readiness_check" {
+  provisioner "local-exec" {
+    when = create
+    command = "kubectl rollout status deployment eks-appmesh-controller -n${kubernetes_namespace.appmesh_ns.metadata[0].name}"
```

Comment:
it's the only way to deploy it? any possibility to use terraform?

Reply:
The AppMesh controller is deployed through helm. `appmesh_readiness_check` waits the controller to be ready, otherwise the injection of envoy sidecar would fail for the following traffic application deployment.